src/tqec/computation/block_graph.py

```diff
@@ -641,7 +641,79 @@ def rotate(
                 pos_map[pipe.v.position],
                 cast(PipeKind, rotated_kind),
             )
-        return rotated
+        return rotated.fix_shadowed_faces()
+
+    def fix_shadowed_faces(self) -> BlockGraph:
+        """Fix the basis of those shadowed faces of the cubes in the graph.
+
+        A pair of face can be shadowed if the cube is connected to two pipes
+        in the same direction. Though these faces are not visible in the 3D
+        visualization, they are still identified by the cube kind and can
+        affect the circuit compilation.
+
+        The basis of the cube is enforced to match the pipes connected to it.
+        If there is unmatched shadowed faces, we try to fix their basis. This
+        will leave some freedom in constructing the model in SketchUp and make
+        everyones' life easier.
+
+        Additionally, for a spatial pass-through, i.e. a cube only connected to
+        two pipes in the same direction, we will fix the cube kind to not be a
+        spatial cube.
+
+        Note that the fixed graph is still not guaranteed to be valid as some
+        other conditions may be violated. You still need to call
+        :py:meth:`~tqec.computation.block_graph.BlockGraph.validate` to check
+        the validity of the graph.
+
+        Returns:
+            A new graph with the shadowed faces fixed.
+        """
+        fixed_cubes: dict[Cube, Cube] = {}
+        for cube in self.cubes:
+            pipes_by_direction: dict[Direction3D, list[Pipe]] = {}
+            for pipe in self.pipes_at(cube.position):
+                pipes_by_direction.setdefault(pipe.direction, []).append(pipe)
+            # Regularize the cube in a spatial pass-through
+            if len(pipes_by_direction) == 1:
+                direction = next(iter(pipes_by_direction))
+                if len(pipes_by_direction[direction]) != 2:
+                    continue
+                if direction == Direction3D.Z or not cube.is_spatial:
+                    continue
+                kind = cube.kind
+                assert isinstance(kind, ZXCube)
+                basis = kind.get_basis_along(direction)
+                new_kind = kind.with_basis_along(direction, basis.flipped())
+                new_cube = Cube(cube.position, new_kind, cube.label)
+                fixed_cubes[cube] = new_cube
+            # Fix the shadowed faces
+            if len(pipes_by_direction) != 2:
+                continue
```

Comment:
If I understood correctly:

- `len(pipes_by_direction) == 0` can only be seen for an isolated cube (memory experiment or stability experiment),
- `len(pipes_by_direction) == 1` is a spatial pass-through that might needs to be corrected,
- `len(pipes_by_direction) == 2` may be a spatial pass-through, that might need to be corrected, or a regular spatial junction, or a regular junction with one pipe in space and one pipe in time,
- `len(pipes_by_direction) == 3` is impossible, because that would make a spatial junction with a pipe in time, which is disallowed.

If that is effectively the case, maybe this `if` can be accompanied with a comment explaining why all the other cases are not concerned by the method.